File: android/src/main/java/expo/modules/stableid/StableIdModule.kt
package expo.modules.stableid

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class StableIdModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoStableId")

    AsyncFunction("fetchAppTransactionId") {
      null as String?
    }
  }
}

File: ios/StableIdModule.swift
import ExpoModulesCore
import StoreKit

public class StableIdModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoStableId")

    AsyncFunction("fetchAppTransactionId") { () -> String? in
      if #available(iOS 16.0, *) {
        do {
          let result = try await AppTransaction.shared
          switch result {
          case .verified(let transaction):
            return String(transaction.appTransactionID)
          case .unverified:
            return nil
          }
        } catch {
          return nil
        }
      }
      return nil
    }
  }
}
